app/src/main/java/com/example/diceroller/DiceFragment.kt

```diff
@@ -1,82 +1,73 @@
+        binding = this
+        this.root
     }
 
-    override fun onActivityCreated(savedInstanceState: Bundle?) {
-        super.onActivityCreated(savedInstanceState)
-        viewModel = ViewModelProvider(requireActivity()).get(RolledDiceViewModel::class.java)
-        subscribeToObservables()
-
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        binding.rollButton.setOnClickListener { viewModel.onRollClicked() }
     }
 
-    private fun subscribeToObservables() {
-        lifecycleScope.launchWhenStarted {
-            viewModel.currentDice.collectLatest {
-
-                val updatedText = "Come on, roll the ${it.name}"
-                binding.resultText.text = updatedText
+    private fun observeState() {
 
-                val diceTypeDrawable = it.diceTypeIcon
-                binding.diceImage.setImageResource(diceTypeDrawable)
-                binding.diceImage.setColorFilter(ContextCompat.getColor(requireContext(), R.color.white))
-
-            }
+        fun handleReadyToRoll(it: ViewState.RollDie.ReadyToRoll) = binding.run {
+            resultText.text = getString(R.string.prompt_roll_die, it.diceType)
+            diceImage.setImageResource(it.diceType.diceTypeIcon)
+            diceImage.setColorFilter(
+                ContextCompat.getColor(requireContext(), R.color.white)
+            )
         }
-    }
-
-    override fun onDestroyView() {
-        super.onDestroyView()
-        _binding = null
-    }
 
-    private fun recieveDice() {
-        val newDice = viewModel.rollDice()
-        val diceText: TextView = binding.resultText
-        val diceTextResult = "You got a ${newDice.result}"
-        diceText.text = diceTextResult
+        fun handleLatestRoll(it: ViewState.RollDie.Rolled) = binding.run {
+            resultText.text =
+                getString(R.string.roll_result_message, it.rollValue)
 
-        val diceImageResult = viewModel.getCurrentDiceType().diceImageList[newDice.result-1]
-        binding.diceImage.setImageResource(diceImageResult)
-        binding.diceImage.setColorFilter(ContextCompat.getColor(requireContext(), com.google.android.material.R.color.mtrl_btn_transparent_bg_color))
+            diceImage.setImageResource(it.diceType.diceImages[it.rollValue - 1])
+            diceImage.setColorFilter(
+                ContextCompat.getColor(requireContext(), R.color.transparent)
+            )
+        }
 
+        lifecycleScope.launchWhenStarted {
+            viewModel.viewState.collect {
+                when (it) {
+                    is ViewState.RollDie.ReadyToRoll -> handleReadyToRoll(it)
+                    is ViewState.RollDie.Rolled -> handleLatestRoll(it)
+                    ViewState.History ->
+                        findNavController().navigate(
+                            DiceFragmentDirections.actionDiceFragmentToRolledDicesListFragment()
+                        )
```

Comment:
This is navigation with SafeArgs. To understand why this is better you have to understand that the original `navigate()` method takes an ID. An ID can be anything, it will accept any ID from any part of the codebase.

```kotlin
findNavController.navigate(R.id.recycler_view)
```

^ This would look like perfectly good code but you can obviously not navigate to a RecyclerView. With SafeArgs you get Direction objects, and those Objects have functions only for place you can navigate to.